functorch/dim/__init__.py

```diff
@@ -168,3 +169,83 @@ def _def(name, *args, **kwargs):
 # scatter_add
 # scatter_add_
 # scatter_reduce
+
+
+def rearrange(tensor: Tensor, equation: str, **kwargs) -> Tensor:
+    r"""Flattens or coalesces dimensions of the input tensor based on the given equation.
+    Implemented as a thin wrapper around `functorch.dim.dims`
+
+    Equation:
+        The equation must be of the form `axes_before -> axes_after`, with the grouping of the
+        axes following the Einstein notation.
+
+        Both sides of the equation will take a series of space-seperated axis groupings.
+        The groupings are positional, and can be either of the form
+        `axis_name` (single axis) or `(axis_0, axis_1, ..)`.
+
+        There must be exactly as many axis groupings on the left as the tensor
+        dimension of the input tensor. Individual axis names must appear exactly once on both sides.
+
+        Where there is more than one axis in a given axis grouping on the left (input side), the
+        dimensions of at most one axis can be unknown (so it can be inferred from the input tensor shape).
+
+        There should not be any spaces separating brackets and axis names.
+
+    Examples:
+        >>> # xdoctest: +IGNORE_WANT("non-deterministic")
+        >>> # simple
+        >>> tensor = torch.rand(1, 4, 2, 2)
+        >>> tensor_out = functorch.dim.rearrange(tensor, 'b c h w -> b h w c')
+        >>> tensor_out.shape
+        torch.Size([1, 2, 2, 4])
+
+        >>> # xdoctest: +IGNORE_WANT("non-deterministic")
+        >>> # with unknown dim kwargs
+        >>> tensor = torch.rand(1, 4, 2, 2)
+        >>> tensor_out = functorch.dim.rearrange(tensor, 'b (c h2 w2) h w -> b c (h h2) (w w2)', h2=2, w2=2)
+        >>> tensor_out.shape
+        torch.Size([1, 1, 4, 4])
+
+    Args:
+        tensor (Tensor): The tensor to rearrange
+        equation (str): The axes and their groupings in Einstein-notation
+        kwargs: arguments of the form `axis_name=size`
+
+    Raises:
+        ValueError/SyntaxError/NameError: If equation is not validly constructed.
+    """
+    # returns lhs, rhs, and axes. Does minor validation of input equation.
+    # the remainder of useful error messages are handled naturally by the generated code.
+    def parse_equation(equation: str) -> Tuple[str, str, str]:
+        split = equation.split('->')
+        if len(split) != 2:
+            raise ValueError(f"While parsing equation: {equation} should be of the form `before_axes -> after_axes`")
+        lhs, rhs = split[0].split(), split[1].split()
+
+        lhs_idents, rhs_idents = set(), set()
+
+        for ident in lhs:
+            lhs_idents.add(ident.replace('(', '').replace(')', ''))
+        for ident in rhs:
+            rhs_idents.add(ident.replace('(', '').replace(')', ''))
```

Comment:
Uhh, is this even right? Don't we have to do something with the parenthesization